handler/oauth2/flow_token_exchange.go

```diff
@@ -0,0 +1,170 @@
+	}
+
+	// subject_token REQUIRED
+	form := request.GetRequestForm()
+	subjectToken := form.Get("subject_token")
+	if subjectToken == "" {
+		return errors.WithStack(fosite.ErrInvalidRequestObject.WithHintf("Mandatory parameter subject_token is missing."))
+	}
+
+	// subject_token_type REQUIRED
+	subjectTokenType := form.Get("subject_token_type")
+	if subjectTokenType == "" {
+		return errors.WithStack(fosite.ErrInvalidRequestObject.WithHintf("Mandatory parameter subject_token_type is missing."))
+	}
+
+	if subjectTokenType != "urn:ietf:params:oauth:token-type:access_token" {
+		return errors.WithStack(fosite.ErrInvalidRequestObject.WithHintf("Currently only subject_token_type urn:ietf:params:oauth:token-type:access_token is supported"))
+	}
+
+	sig := c.CoreStrategy.AccessTokenSignature(subjectToken)
+	or, err := c.CoreStorage.GetAccessTokenSession(ctx, sig, request.GetSession())
+	if err != nil {
+		return errors.WithStack(fosite.ErrRequestUnauthorized.WithDebug(err.Error()))
+	} else if err := c.CoreStrategy.ValidateAccessToken(ctx, or, subjectToken); err != nil {
+		return err
+	}
+
+	var delegatingClient fosite.Client = nil
+	// reload client from storage to ensure that may_act is up to date in case of eventual revocation
+	if or.GetDelegatingClient() == nil {
+		// first exchange request has no delegating client set
+		delegatingClient, err = c.Store.GetClient(ctx, or.GetClient().GetID())
+	} else {
+		delegatingClient, err = c.Store.GetClient(ctx, or.GetDelegatingClient().GetID())
+	}
+
+	if err != nil {
+		return errors.WithStack(fosite.ErrInvalidClient.WithHint("The delegating OAuth2 Client does not exist."))
+	}
+
+	// check if delegating client allows the current client to perform an exchange on its tokens
+	if !delegatingClient.GetMayAct().HasOneOf(client.GetID()) {
+		return errors.WithStack(fosite.ErrUnauthorizedClient.WithHint("The OAuth 2.0 Client is not allowed to perform a token exchange for the given subject token."))
+	}
+	request.SetDelegatingClient(delegatingClient)
+
+	for _, scope := range request.GetRequestedScopes() {
+		if !c.ScopeStrategy(client.GetScopes(), scope) &&
+			!c.ScopeStrategy(or.GetGrantedScopes(), scope) {
+			return errors.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope \"%s\".", scope))
+		}
+	}
+
+	if err := c.AudienceMatchingStrategy(client.GetAudience(), request.GetRequestedAudience()); err != nil {
+		return err
+	}
+
+	if client.IsPublic() {
+		return errors.WithStack(fosite.ErrInvalidGrant.WithHint("The OAuth 2.0 Client is marked as public and is thus not allowed to use authorization grant \"urn:ietf:params:oauth:grant-type:token-exchange\"."))
+	}
+
+	request.GetSession().SetExpiresAt(fosite.AccessToken, time.Now().UTC().Add(c.AccessTokenLifespan))
+	if c.RefreshTokenLifespan > -1 {
+		request.GetSession().SetExpiresAt(fosite.RefreshToken, time.Now().UTC().Add(c.RefreshTokenLifespan).Round(time.Second))
+	}
```

Comment:
I assume this is needed because `RefreshTokenGrantHandler` doesn't handle this `grant_type`? However, to me it also looks like `requested_token_type` should be responsible for setting this.

Reply:
The logic we use here is pretty much the same as the the one already used e.g. `flow_authorize_code_token`.
I disagree, regarding the use of `requested_token_type`.

> requested_token_type
OPTIONAL . An identifier, as described in Section 3, for the type of the requested security
token. ...

In my understanding, a refresh token is not really a security token. I also do not see the point of making an exchange request just to get an refresh token.
Perhaps I miss understood you?